repo/open.go

```diff
@@ -183,51 +186,125 @@ func openDirect(ctx context.Context, configFile string, lc *LocalConfig, passwor
+	err error,
+) {
+	// Read format blob, potentially from cache.
+	fb, cacheTime, err = readAndCacheRepositoryBlobBytes(ctx, st, cacheOpts.CacheDirectory, FormatBlobID, validDuration)
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to read repository blobs")
+		return nil, time.Time{}, nil, nil, nil, errors.Wrap(err, "unable to read format blob")
 	}
 
-	if err = writeCacheMarker(caching.CacheDirectory); err != nil {
-		return nil, errors.Wrap(err, "unable to write cache directory marker")
+	if err = writeCacheMarker(cacheOpts.CacheDirectory); err != nil {
+		return nil, time.Time{}, nil, nil, nil, errors.Wrap(err, "unable to write cache directory marker")
 	}
 
-	f, err := parseFormatBlob(fb)
+	f, err = parseFormatBlob(fb)
 	if err != nil {
-		return nil, errors.Wrap(err, "can't parse format blob")
+		return nil, time.Time{}, nil, nil, nil, errors.Wrap(err, "can't parse format blob")
 	}
 
 	fb, err = addFormatBlobChecksumAndLength(fb)
 	if err != nil {
-		return nil, errors.Errorf("unable to add checksum")
+		return nil, time.Time{}, nil, nil, nil, errors.Errorf("unable to add checksum")
 	}
 
-	formatEncryptionKey, err := f.deriveFormatEncryptionKeyFromPassword(password)
+	formatEncryptionKey, err = f.deriveFormatEncryptionKeyFromPassword(password)
 	if err != nil {
-		return nil, err
+		return nil, time.Time{}, nil, nil, nil, err
 	}
 
-	repoConfig, err := f.decryptFormatBytes(formatEncryptionKey)
+	repoConfig, err = f.decryptFormatBytes(formatEncryptionKey)
 	if err != nil {
-		return nil, ErrInvalidPassword
+		return nil, time.Time{}, nil, nil, nil, ErrInvalidPassword
+	}
+
+	return fb, cacheTime, f, repoConfig, formatEncryptionKey, nil
+}
+
+// ReadAndCacheRepoConfig loads the config from cache and returns the cache
+// time and the repo-config object.
+func ReadAndCacheRepoConfig(
+	ctx context.Context,
+	st blob.Storage,
+	password string,
+	cacheOpts *content.CachingOptions,
+	validDuration time.Duration,
+) (
+	cacheTime time.Time,
+	repoConfig *repositoryObjectFormat,
+	err error,
+) {
+	// nolint:dogsled
+	_, cacheTime, _, repoConfig, _, err = readAndCacheRepoConfig(ctx, st, password, cacheOpts, validDuration)
+	return cacheTime, repoConfig, err
+}
+
+// openWithConfig opens the repository with a given configuration, avoiding the need for a config file.
+// nolint:funlen,gocyclo
+func openWithConfig(ctx context.Context, st blob.Storage, lc *LocalConfig, password string, options *Options, cacheOpts *content.CachingOptions, configFile string) (DirectRepository, error) {
+	cacheOpts = cacheOpts.CloneOrDefault()
+	cmOpts := &content.ManagerOptions{
+		TimeNow:            defaultTime(options.TimeNowFunc),
+		DisableInternalLog: options.DisableInternalLog,
+	}
+
+	var (
+		cacheTime           time.Time
+		f                   *formatBlob
+		repoConfig          *repositoryObjectFormat
+		formatEncryptionKey []byte
+	)
+
+	if err := retry.WithExponentialBackoffNoValue(ctx, "read repo config and wait for upgrade", func() error {
```

Comment:
btw, as of now this exponential backoff is limited and very short overall duration, just a few seconds (it's a bug, I think)

you want a back-off here that lasts forever-ish, which we don't have a helper for.